subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/AbstractVirtualFileSystem.java

```diff
@@ -22,17 +22,24 @@
 import org.gradle.internal.vfs.VirtualFileSystem;
 
 import java.util.Optional;
+import java.util.function.Function;
 
 public abstract class AbstractVirtualFileSystem implements VirtualFileSystem {
 
+    protected final VfsRootReference rootReference;
```

Comment:
This should only be visible for testing, right?

Reply:
`WatchingVirtualFileSystem` uses the `rootReference` in various places when it updates the root without notifying itself.